### PR TITLE
[FEATURE] Git blacklist formatter

### PR DESCRIPTION
**Pull request by Kanti**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documented?   | -
| Fixed tickets | -

Nicer Git Blacklist Task message for very long lines.
Example:
image

### CODE REVIEW

**Comment by veewee** [comment]
Thanks for the PR @Kanti! At first sight, it looks like a great formatter.
This formatter requires ANSI escape codes but there are users on e.g. who use terminals which don't support those colors. Can you check `IOInterface::isDecorated()` to enable / disable colored output?

I'll try to dive in the code somewhere this week.

**Comment by Kanti** [comment]
I have found something strange. 
I have added This:
`````php
var_dump(get_class($container->get('grumphp.io')));
````
in this line https://github.com/phpro/grumphp/blob/master/src/GrumPHP/Console/Application.php#L178

There I should get ``string(20) "GrumPHP\IO\ConsoleIO"`` printed.
But instead I get ``string(17) "GrumPHP\IO\NullIO"`` printed.

Is this a Bug? Could you fix it?

**Comment by veewee** [comment]
That indeed is something strange. I'll have to dive into that one.

**Comment by veewee** [comment]
Hi @Kanti,

Can you tell me if this is still an issue if you mark the task as a lazy service?

```yml
services:
    formatter.git_blacklist:
        class: GrumPHP\Formatter\GitBlacklistFormatter
        lazy: true
        arguments:
            - '@grumphp.io'
```

Currently it is possible that the `TaskRunner` is initialized before the `configureIO` method is being called. This is indeed a problem which we need to fix. In the meantime, the `lazy` option can be a quick fix.

File: src/GrumPHP/Formatter/GitBlacklistFormatter.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            } else {
+                $result .= $line . PHP_EOL;
+            }
+        }
+        return trim($result);
+    }
+
+    private function trimOutputLine($line, $lineNumber)
+    {
+        $wordColor = "\033[1;31";
```
Comment (veewee) [inline src/GrumPHP/Formatter/GitBlacklistFormatter.php]: Maybe these colors should be class constants.

File: src/GrumPHP/Formatter/GitBlacklistFormatter.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            if (preg_match('/^[0-9]+/', $line)) {
+                $result .= $this->trimOutputLine($line, $lineNumber) . PHP_EOL;
+            } else {
+                $result .= $line . PHP_EOL;
+            }
+        }
+        return trim($result);
+    }
+
+    private function trimOutputLine($line, $lineNumber)
```
Comment (veewee) [inline src/GrumPHP/Formatter/GitBlacklistFormatter.php]: Missing method docblocks.

File: src/GrumPHP/Formatter/GitBlacklistFormatter.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        $wordColor = "\033[1;31";
+        $resetColor = "\033[m";
+        $colonColor = "\033[36m";
+        $coloredColon = $colonColor . ':' . $resetColor;
+
+        $before = 20;
+        $after = 20;
+
+        if (strlen($line) >= 80) {
```
Comment (veewee) [inline src/GrumPHP/Formatter/GitBlacklistFormatter.php]: It's more readable if you invert the if statement and add an early return at this point.
You might want to split the logic in this if into a new private method.

File: src/GrumPHP/Formatter/GitBlacklistFormatter.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            $lastPos = 0;
+            $positionsFirst = array();
+            $positionsSecond = array();
+            $parts = array();
+
+            while (($lastPos = mb_strpos($line, $wordColor, $lastPos)) !== false) {
```
Comment (veewee) [inline src/GrumPHP/Formatter/GitBlacklistFormatter.php]: There are a lot of `while` iterations going on in here. Can you add some comments above these iterations about what those iterations are doing? Are there infinite loops possible?

File: src/GrumPHP/Task/Git/Blacklist.php [hunk 5]
```diff
@@ -16,6 +20,22 @@
 class Blacklist extends AbstractExternalTask
 {
     /**
+     * @var IOInterface
+     */
+    private $IO;
+
+    public function __construct(
```
Comment (veewee) [inline src/GrumPHP/Task/Git/Blacklist.php]: Missing docblocks

File: src/GrumPHP/Formatter/GitBlacklistFormatter.php (hunk 4 above)
Comment (Kanti) [inline src/GrumPHP/Formatter/GitBlacklistFormatter.php]: Are there infinite loops possible?
=> no, $lastPos will get bigger on every iteration.

**Comment by Kanti** [comment]
Hi @veewee can you look into the bug in the next days?

**Comment by veewee** [comment]
Hi @Kanti,

I'll try to pick it up somewhere this week. Sorry for the latency :)

**Comment by veewee** [comment]
Hi @Kanti,

I've added late input / output binding to the console IO.
Can you verify if this works for you? It seems to work here.

**Comment by Kanti** [comment]
all good from my side. 😃

**Pull request merged**